website/pages/docs/internals/security.mdx

```diff
@@ -0,0 +1,322 @@
+    restricted to specific users.
+
+* **[Sentinel Policies](/guides/governance-and-policy/sentinel/sentinel-policy.html)**
+    (**Enterprise Only**) - [Sentinel](https://www.hashicorp.com/sentinel/) is
+    a feature which enables
+    [policy-as-code](https://docs.hashicorp.com/sentinel/concepts/policy-as-code/)
+    to enforce further restrictions on operators. This is used to augment the
+    built-in ACL system for fine-grained control over jobs.
+
+* **[Namespaces](/guides/governance-and-policy/namespaces.html)**
+    (**Enterprise Only**) - This feature allows for a cluster to be shared by
+    multiple teams within a company. Using this logical separation is important
+    for multi-tenant clusters to prevent users without access to that namespace
+    from conflicting with each other. This requires ACLs to be enabled in order
+    to be enforced.
+
+* **[Resource Quotas](/guides/governance-and-policy/quotas.html)**
+    (**Enterprise Only**) - Can limit a namespace’s access to the underlying
+    compute resources in the cluster by setting upper-limits for operators.
+    Access to these resource quotas can be managed via ACLs to ensure read-only
+    access for operators so they can’t just change their quotas.
+
+#### Recommendations
+
+The following are security recommendations that can help significantly improve
+the security of your cluster depending on your use case. We recommend always
+practicing defense in depth when architecting the security mechanisms for your
+environment.
+
+* **[Rotate Credentials](/docs/job-specification/vault.html)** -
+    Using something like [Vault](/docs/vault-integration/index.html) to
+    create and manage dynamic, rotated credentials is highly recommended to
+    prevent secrets from being easily exposed within the [job
+    specification](/docs/job-specification/index.html)
+    itself which may be leaked into version control or otherwise be accidently
+    stored on disk on an operator’s local machine. It is also possible to
+    [integrate with Vault’s PKI secret engine](/guides/security/vault-pki-integration.html)
+    to automatically generate and renew dynamic, unique X.509 certificates for
+    each Nomad node with a short
+    [TTL](https://en.wikipedia.org/wiki/Time_to_live).
+
+* **[Running without Root](https://groups.google.com/forum/#!topic/nomad-tool/pSyMwC_FSFA)** -
+    Certain features of Nomad can be used without needing to run the Nomad agent
+    server or client as the `root` user. Instead you can granularly assign the
+    appropriate capabilities in various ways for your Nomad agents. For example:
+    Nomad servers only require access to the data directory; it is possible to
+    use Nomad to orchestrate Docker containers by adding a non-root `nomad` user
+    to the `docker` group to access the [default unix
+    socket](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option).
```

Comment:
Making this a "recommendation" makes me nervous as there are tradeoffs involved. As specified in the linked message, running as unprivileged user prevents Nomad from using an in-memory filesystem for the per-task `/secrets` directory which causes all secrets to be persisted to disk.

Likewise running `raw_exec` with an unprivileged Nomad agent is actually _less_ safe than running `raw_exec` with a root Nomad agent. Since `raw_exec` tasks can only run as the same user as Nomad when Nomad is not root, they have full access to Nomad's data and configuration directories which may contain secrets that would not be divulged otherwise.

That being said: running `raw_exec` is *never* secure, so perhaps that example isn't valid.

I think running Nomad as non-root is only useful for functionally reducing jobs' privilege level to Docker's and therefore easing threat modelling and hardening in Docker-only environments.

Podman's rootless containers may make this a stronger recommendation since they provide task isolation while avoiding executing any privileged code (unlike dockerd). Thus simplifying the threat model by always providing tasks a _subset_ of Nomad's capabilities.

Again I'm fine with publishing this as-is, since I think it's even argue to pros of a non-root Nomad outweigh the cons.

Reply:
I'll restrict this recommendation to servers. It's definitely hard to run clients without being root.

Reply:
Oh that's perfect. There's no reason for servers to run as root, and I don't think we make that clear anywhere.